From: Ralf Baechle
Date: Sat, 5 Dec 2009 20:19:59 +0000 (+0000)
Subject: MIPS: Fix potencial aliases due to broken check in arch_get_unmapped_area
X-Git-Url: http://www.linux-mips.org/git?p=linux.git;a=commitdiff_plain;h=860f61209111af27eacfbe77a29b0e70bdd93a72

MIPS: Fix potencial aliases due to broken check in arch_get_unmapped_area

Noticed by Al Viro <viro@ftp.linux.org.uk>.

Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
(cherry picked from commit 1d8a38bc5ec86a44d1eae863099f51766735d716)
---

diff --git a/arch/mips/kernel/syscall.c b/arch/mips/kernel/syscall.c
index 505358b..fe26fce 100644
--- a/arch/mips/kernel/syscall.c
+++ b/arch/mips/kernel/syscall.c
@@ -72,7 +72,8 @@ unsigned long arch_get_unmapped_area(struct file *filp, unsigned long addr,
 		 * We do not accept a shared mapping if it would violate
 		 * cache aliasing constraints.
 		 */
-		if ((flags & MAP_SHARED) && (addr & shm_align_mask))
+		if ((flags & MAP_SHARED) &&
+		    ((addr - (pgoff << PAGE_SHIFT)) & shm_align_mask))
 			return -EINVAL;
 		return addr;
 	}
